evi-python-function-calling/main.py

```diff
@@ -0,0 +1,391 @@
+        elif message.type == "error":
+            error_message: str = message.message
+            error_code: str = message.code
+            raise ApiError(f"Error ({error_code}): {error_message}")
+        else:
+            message_type = message.type.upper()
+            text = f"<{message_type}>"
+        
+        # Print the formatted message
+        self._print_prompt(text)
+
+        # Extract and print the top 3 emotions inferred from user and assistant expressions
+        if len(scores) > 0:
+            top_3_emotions = self._extract_top_n_emotions(scores, 3)
+            self._print_emotion_scores(top_3_emotions)
+            print("")
+        else:
+            print("")
+        
+    async def on_close(self):
+        """Logic invoked when the WebSocket connection is closed."""
+        print("WebSocket connection closed.")
+
+    async def on_error(self, error):
+        """Logic invoked when an error occurs in the WebSocket connection.
+        
+        See the full list of errors [here](https://dev.hume.ai/docs/resources/errors).
+
+        Args:
+            error (Exception): The error that occurred during the WebSocket communication.
+        """
+        print(f"Error: {error}")
+
+    def _print_prompt(self, text: str) -> None:
+        """Print a formatted message with a timestamp.
+
+        Args:
+            text (str): The message text to be printed.
+        """
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        now_str = now.strftime("%H:%M:%S")
+        print(f"[{now_str}] {text}")
+
+    def _extract_top_n_emotions(self, emotion_scores: dict, n: int) -> dict:
+        """
+        Extract the top N emotions based on confidence scores.
+
+        Args:
+            emotion_scores (dict): A dictionary of emotions and their corresponding confidence scores.
+            n (int): The number of top emotions to extract.
+
+        Returns:
+            dict: A dictionary containing the top N emotions as keys and their raw scores as values.
+        """
+        # Convert the dictionary into a list of tuples and sort by the score in descending order
+        sorted_emotions = sorted(emotion_scores.items(), key=lambda item: item[1], reverse=True)
+
+        # Extract the top N emotions
+        top_n_emotions = {emotion: score for emotion, score in sorted_emotions[:n]}
+
+        return top_n_emotions
+
+    def _print_emotion_scores(self, emotion_scores: dict) -> None:
+        """
+        Print the emotions and their scores in a formatted, single-line manner.
+
+        Args:
+            emotion_scores (dict): A dictionary of emotions and their corresponding confidence scores.
+        """
+        # Format the output string
+        formatted_emotions = ' | '.join([f"{emotion} ({score:.2f})" for emotion, score in emotion_scores.items()])
+        
+        # Print the formatted string
+        print(f"|{formatted_emotions}|")
+
+async def fetch_weather(location: str, format: str) -> str:
```

Comment:
@norman-codes Let's pull this util and other utils, which are not specific to managing the socket, out to another file to clean things up a bit.

Reply:
Will do!